zebra-state/src/service.rs

```diff
@@ -252,6 +252,87 @@ impl StateService {
     pub fn utxo(&self, outpoint: &transparent::OutPoint) -> Option<transparent::Output> {
         self.mem.utxo(outpoint).or_else(|| self.sled.utxo(outpoint))
     }
+
+    /// Return an iterator over the relevant chain of the block identified by
+    /// `hash_or_height`.
+    #[allow(dead_code)]
+    pub fn chain(&self, hash_or_height: HashOrHeight) -> Iter<'_> {
+        Iter {
+            service: self,
+            state: State::NonFinalized(hash_or_height),
+        }
+    }
+}
+
+struct Iter<'a> {
+    service: &'a StateService,
+    state: State,
+}
+
+enum State {
+    NonFinalized(HashOrHeight),
+    Finalized(block::Height),
+    Finished,
+}
+
+impl<'a> Iter<'a> {
+    fn next_non_finalized_block(&mut self) -> Option<Arc<Block>> {
+        let Iter { service, state } = self;
+
+        let hash_or_height = match state {
+            State::NonFinalized(hash_or_height) => hash_or_height,
+            State::Finalized(_) | State::Finished => unreachable!(),
+        };
+
+        if let Some(block) = service.mem.block(*hash_or_height) {
+            let hash = block.header.previous_block_hash;
+            self.state = State::NonFinalized(hash.into());
+            Some(block)
+        } else {
+            None
+        }
+    }
+
+    fn next_finalized_block(&mut self) -> Option<Arc<Block>> {
+        let Iter { service, state } = self;
+
+        let hash_or_height = match *state {
+            State::Finalized(height) => height.into(),
+            State::NonFinalized(hash_or_height) => hash_or_height,
+            State::Finished => unreachable!(),
+        };
+
+        if let Some(block) = service.sled.block(hash_or_height) {
+            let height = block
+                .coinbase_height()
+                .expect("valid blocks have a coinbase height");
+
+            if let Some(next_height) = height - 1 {
+                self.state = State::Finalized(next_height);
+            } else {
+                self.state = State::Finished;
+            }
+
+            Some(block)
+        } else {
+            self.state = State::Finished;
+            None
+        }
+    }
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = Arc<Block>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.state {
+            State::NonFinalized(_) => self
+                .next_non_finalized_block()
+                .or_else(|| self.next_finalized_block()),
+            State::Finalized(_) => self.next_finalized_block(),
+            State::Finished => None,
+        }
+    }
 }
 
```

Comment:
Let's also implement:
* `Iterator::size_hint`
  * return the height of the hash in the iterator's state
  * https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.size_hint
* `ExactSizeIterator`:
  * the default `len` implementation is good
  * https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html
* `FusedIterator`
  * https://doc.rust-lang.org/std/iter/trait.FusedIterator.html

I don't think there's any reason to implement `DoubleEndedIterator` at this point.

Reply:
Since this is an entirely internal API I'd prefer to delay implementing them until we have a concrete use case.

Reply:
`Iterator::size_hint` and `ExactSizeIterator` allow us to efficiently assert that the chain is at least 28 blocks long.

From the `FusedIterator` docs:
> Calling next on a fused iterator that has returned None once is guaranteed to return None again. This trait should be implemented by all iterators that behave this way because it allows optimizing Iterator::fuse.

There's no particular reason to implement `FusedIterator` right now, but it also doesn't cost us anything. `FusedIterator` has no methods - it's only used to make `Iterator::fuse` into a no-op.

`fuse` is used internally by `flatten` and possibly other iterator methods. So I don't think we'd necessarily notice when we started using `fuse`, and even if we did, we might not remember to implement `FusedIterator` on the `chain` iterator.

Reply:
> `Iterator::size_hint` and `ExactSizeIterator` allow us to efficiently assert that the chain is at least 28 blocks long.

This property isn't particularly useful though, because the caller `check::block_difficulty` or w/e its called will have access to the block itself, and will know how long its chain would be. Also, iirc we put in the RFC that we assume that there are at least 28 blocks of finalized state whenever we call `check::block_difficulty`, so we don't need to assert this in advanced, just panic if we find we have insufficient blocks while iterating.